### PR TITLE
Adding deterministic data-live-id attribute to child components

### PR DESCRIPTION
**Pull request by weaverryan**
> 4) BACKEND: pass the a deterministic data-live-id attribute to the child unless manually specified

From https://github.com/symfony/ux/pull/500

A little wild, but we need a way to be able to identify child components from each other when a parent component is re-rendering. Livewire also does this.

The flow is:

A) Component parent renders & all children are given a "fingerprint" (still a TODO - 5 on the linked PR) that is a checksum of the "input" passed to create that child
B) On parent re-render, an array containing `[{ id: 'the-child-id', fingerprint: 'its fingerprint'}, ... ]` is sent along with the parent component data.
C) We will (still a TODO - 6 on the linked PR) use that info to figure out if the "fingerprint" of any specific children has changed (i.e. we calculate the deterministic id of a child component, find its fingerprint using that id from the data above, and compare).

### CODE REVIEW

File: src/TwigComponent/src/ComponentRenderer.php [hunk 1]
```diff
@@ -44,9 +45,14 @@ public function createAndRender(string $name, array $props = []): string
...
     {
+        $this->componentStack->push($mounted);
+
         $event = $this->preRender($mounted);
 
-        return $this->twig->render($event->getTemplate(), $event->getVariables());
+        $rendered = $this->twig->render($event->getTemplate(), $event->getVariables());
+        $this->componentStack->pop();
+
+        return $rendered;
```
Comment (kbond) [inline src/TwigComponent/src/ComponentRenderer.php]:
```suggestion
        try {
            return $this->twig->render($event->getTemplate(), $event->getVariables());
        } finally {
            $this->componentStack->pop();
        }
```

File: src/TwigComponent/src/ComponentRenderer.php [hunk 2]
```diff
@@ -44,9 +45,14 @@ public function createAndRender(string $name, array $props = []): string
 
     public function render(MountedComponent $mounted): string
```
Comment (kbond) [inline src/TwigComponent/src/ComponentRenderer.php:46]: So the stack will always be empty after the "primary render" is finished?

File: src/LiveComponent/src/Twig/DeterministicIdCalculator.php [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+ *
+ * @experimental
+ *
+ * @internal
+ */
+class DeterministicIdCalculator
+{
+    private array $childRenderCounts = [];
+
+    public function calculateDeterministicId(): string
```
Comment (kbond) [inline src/LiveComponent/src/Twig/DeterministicIdCalculator.php]: Bowser Wowser!

Can you add some comments here to explain what's happening?

File: src/LiveComponent/src/EventListener/AddLiveAttributesSubscriber.php [hunk 4]
```diff
@@ -100,6 +107,20 @@ private function getLiveAttributes(MountedComponent $mounted, ComponentMetadata
...
+        if ($this->hasParentComponent()) {
+            $id = $this->container->get(DeterministicIdCalculator::class)->calculateDeterministicId();
+
+            $attributes['data-live-id'] = $id;
+        }
+
         return new ComponentAttributes($attributes);
     }
+
+    private function hasParentComponent(): bool
```
Comment (kbond) [inline src/LiveComponent/src/EventListener/AddLiveAttributesSubscriber.php]: Thinking this method should exist on `ComponentStack`.

File: src/TwigComponent/src/ComponentRenderer.php (hunk 2 above)
Comment (weaverryan) [reply src/TwigComponent/src/ComponentRenderer.php:46]: Yup, that's how I designed it. No reason, at the moment, to need to know which components have rendered earlier in this request, for example.

**Pull request merged**